src/QuadForm/indefiniteLLL.jl

```diff
@@ -0,0 +1,357 @@
+The first $rank$ columns of $U$ span the kernel.; check::Bool = false
+=#
+function _ker_mod_p(M::MatElem,p)
+  R = parent(M[1,1])
+  rk, k = kernel(change_base_ring(ResidueRing(R,p),M))
+  U = _complete_to_basis(lift(k[:,1:rk]))
+  reverse_cols!(U)
+
+  return rk, U
+end
+
+################################################################################
+#                           Quadratic Equations
+################################################################################
+
+#=
+    _quad_form_solve_triv(G::MatElem{fmpz}; base::Bool = false) -> MatElem{fmpz}
+
+Trying to solve $G$ = 0 with small coefficients. Works if $det$($G$) = 1, dim <= 6
+and $G$ is LLL-reduced. Return [$G$,$I_n$] if no solution is found.
+Exit with a norm 0 vector if one such is found.
+If base = 1 and a norm 0 vector is obtained, returns $transpose$($H$)*$G$*$H$, 
+$H$, $sol$ where $sol$ is of norm 0 vand is the first column of $H$.
+=#
+function _quad_form_solve_triv(G::MatElem{fmpz}, base::Bool = false)
+  n = ncols(G)
+  H = one(parent(G))
+
+  #Case 1: A basis vector is isotropic
+  for i = 1:n
+    if(G[i,i] == 0)
+      sol = H[:,i]
+      if(base == false)
+        return sol
+      end
+      H[:,i] = H[:,1]
+      H[:,1] = sol
+
+      return transpose(H)*G*H, H, sol
+    end
+  end
+
+  #Case 2: G has a block +- [1 0 ; 0 -1] on the diagonal
+  for i = 2:n
+    if(G[i-1,i] == 0 && G[i-1,i-1]*G[i,i] == -1)
+  
+      H[i-1,i] = -1
+      sol = H[:,i]
+      if (base == false)
+        return sol
+      end
+      H[:,i] = H[:,1]
+      H[:,1] = sol
+        return transpose(H)*G*H, H, sol
+    end
+  end
+
+  #Case 3: a principal minor is 0
+  for i = 1:n
+    GG = G[1:i,1:i]
+    if(det(GG) != 0)
+      continue
+    end
+    sol = kernel(GG)[2][:,1]
+    sol = divexact(sol,content(sol))
+    sol = vcat(sol,zero_matrix(base_ring(sol),n-i,1))
+    if (base == false)
+      return sol
+    end
+    H = _complete_to_basis(sol)
+    H[:,n] = - H[:,1]
+    H[:,1] = sol
+    
+    return transpose(H)*G*H, H, sol
+  end
+
+  return G,H
+end
+
+################################################################################
+#                           Quadratic Forms Reduction
+################################################################################
+
+@doc Markdown.doc"""
+    quadratic_form_lll_gram_indef(G::MatElem{fmpz}, base=0) 
+                                              -> Tuple{MatElem{fmpz}, MatElem{fmpz}}
+
+Given a Gram matrix `G` of an indefinite quadratic lattice with $det(G) != 0$.
+Return the LLL-reduction of `G` or find an isotropic vector.
+"""
+function quad_form_lll_gram_indef(G::MatElem{fmpz}, base::Bool = false)
+  n = ncols(G)
+  M = one(parent(G))
+  QD = G
+  MS = identity_matrix(QQ,n)
+
+  # GSO breaks off if one of the basis vectors is isotropic
+  for i = 1:n-1
+    if(QD[i,i] == 0)
+      return _quad_form_solve_triv(G,base)
+    end
+
+    M1 = one(MS)
```

Comment:
```suggestion
    M1 = identity_matrix(QQ, n)
```
You defined `MS` only to create copy of it all along. You should discard the definition of `MS` and re-create `M1` in this way at each iteration.